src/Datadog.Trace.ClrProfiler.Managed/Integrations/RabbitMQIntegration.cs

```diff
@@ -0,0 +1,910 @@
+                    instanceType: modelType.AssemblyQualifiedName);
+                throw;
+            }
+
+            RabbitMQTags tags = null;
+            using (var scope = CreateScope(Tracer.Instance, out tags, command, queue: queue))
+            {
+                try
+                {
+                    instrumentedMethod(model, queue, passive, durable, exclusive, autoDelete, nowait, arguments);
+                }
+                catch (Exception ex)
+                {
+                    scope?.Span.SetException(ex);
+                    throw;
+                }
+            }
+        }
+
+        internal static Scope CreateScope(Tracer tracer, out RabbitMQTags tags, string command, ISpanContext parentContext = null, DateTimeOffset? startTime = null, string queue = null, string exchange = null, string routingKey = null)
+        {
+            tags = null;
+
+            if (!tracer.Settings.IsIntegrationEnabled(IntegrationId))
+            {
+                // integration disabled, don't create a scope, skip this trace
+                return null;
+            }
+
+            Scope scope = null;
+
+            try
+            {
+                Span parent = tracer.ActiveScope?.Span;
+
+                tags = new RabbitMQTags();
+                scope = tracer.StartActiveWithTags(OperationName, parent: parentContext, tags: tags, serviceName: $"{tracer.DefaultServiceName}-{IntegrationName}", startTime: startTime);
+                var span = scope.Span;
+
+                span.Type = SpanTypes.Queue;
+                span.ResourceName = command;
+                tags.Command = command;
+
+                tags.Queue = queue;
+                tags.Exchange = exchange;
+                tags.RoutingKey = routingKey;
+
+                tags.InstrumentationName = IntegrationName;
+                tags.SetAnalyticsSampleRate(IntegrationId, tracer.Settings, enabledWithGlobalSetting: false);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, "Error creating or populating scope.");
+            }
+
+            // always returns the scope, even if it's null because we couldn't create it,
+            // or we couldn't populate it completely (some tags is better than no tags)
+            return scope;
+        }
+
+        /********************
+         * Duck Typing Types
+         */
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+#pragma warning disable SA1201 // Elements must appear in the correct order
+#pragma warning disable SA1600 // Elements must be documented
+        public interface IBasicGetResult
+        {
+            /// <summary>
+            /// Gets the message body of the result
+            /// </summary>
+            IBody Body { get; }
+
+            /// <summary>
+            /// Gets the message properties
+            /// </summary>
+            IBasicProperties BasicProperties { get; }
+        }
```

Comment:
If we are just retrieving values from properties and fields, you can do an optimization by using an struct with the `[DuckCopy]` attribute, with that we avoid the extra boxing of the struct proxy to an `IBasicGetResult` when using the `As<IBasicGetResult>` extension method.

You can find an example of this [here](https://github.com/DataDog/dd-trace-dotnet/blob/master/src/Datadog.Trace.ClrProfiler.Managed/Integrations/HttpMessageHandlerIntegration.cs#L328-L351)

In case you need to use the setters or calling a method, you must keep using an interface. Eg: `IBasicProperties`

Reply:
I need to set `IBasicProperties.Headers` so I can't make `IBasicProperties` a struct. But it should be safe to make `IBasicGetResult` a struct right?

Reply:
Also, the `BasicGetResult` may be null. With the appropriate null checks, do you think it will be safe to use a `[DuckCopy]` struct?

Reply:
Thanks for the suggestion, implemented in ddcbedd2e12553d22ffd7b4e434210b8209f566c

Reply:
Sorry I've just saw the replies. But yeah, should work, I guess you already test it. 😅